File: java/parser/src/main/java/gov/lbl/parser/parser/Function_argumentVisitor.java
package gov.lbl.parser.parser;

import gov.lbl.antlr4.visitor.modelicaBaseVisitor;
import gov.lbl.antlr4.visitor.modelicaParser;
import gov.lbl.parser.domain.Expression;
import gov.lbl.parser.domain.Function_argument;
import gov.lbl.parser.domain.Name;
import gov.lbl.parser.domain.Named_arguments;

public class Function_argumentVisitor extends modelicaBaseVisitor<Function_argument> {
    @Override
    public Function_argument visitFunction_argument(modelicaParser.Function_argumentContext ctx) {
        NameVisitor nameVisitor = new NameVisitor();
        Name function_name = ctx.name() == null? null : ctx.name().accept(nameVisitor);

        Named_argumentsVisitor named_argumentsVisitor = new Named_argumentsVisitor();
        Named_arguments named_arguments = ctx.named_arguments() == null ? null : ctx.named_arguments().accept(named_argumentsVisitor);

        ExpressionVisitor expressionVisitor = new ExpressionVisitor();
        Expression expression = ctx.expression() == null ? null : ctx.expression().accept(expressionVisitor);

        return new Function_argument(function_name, named_arguments, expression);
    }
}

File: jsParser/parser/Function_argumentVisitor.js
const modelicaVisitor = require('../antlrFiles/modelicaVisitor').modelicaVisitor
const Function_argument = require('../domain/Function_argument')

const NameVisitor = require('./NameVisitor')
const Named_argumentsVisitor = require('./Named_argumentsVisitor')
const ExpressionVisitor = require('./ExpressionVisitor')

class Function_argumentVisitor {
  constructor () {
    modelicaVisitor.call(this)
    return this
  }

  visitFunction_argument (ctx) {
    let function_name = null
    let named_arguments = null
    let expression = null

    if (ctx.name()) {
      const nameVisitor = new NameVisitor.NameVisitor()
      function_name = nameVisitor.visitName(ctx.name())
    }
    if (ctx.named_arguments()) {
      const named_argumentsVisitor = new Named_argumentsVisitor.Named_argumentsVisitor()
      named_arguments = named_argumentsVisitor.visitNamed_arguments(ctx.named_arguments())
    }
    if (ctx.expression()) {
      const expressionVisitor = new ExpressionVisitor.ExpressionVisitor()
      expression = expressionVisitor.visitExpression(ctx.expression())
    }

    return new Function_argument.Function_argument(function_name, named_arguments, expression)
  }
};

Function_argumentVisitor.prototype = Object.create(modelicaVisitor.prototype)

exports.visitFunction_argument = this.visitFunction_argument
exports.Function_argumentVisitor = Function_argumentVisitor
